contracts/token/CRC1155/extensions/CRC1155Enumerable.sol

```diff
@@ -0,0 +1,112 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import "./ICRC1155Enumerable.sol";
+import "@openzeppelin/contracts/token/ERC1155/ERC1155.sol";
+import "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+
+abstract contract CRC1155Enumerable is ERC1155, ICRC1155Enumerable {
+    using EnumerableSet for EnumerableSet.UintSet;
+
+    // All token ids for enumeration.
+    EnumerableSet.UintSet private _allTokens;
+
+    // Mapping from token ID to token total supply.
+    mapping(uint256 => uint256) private _totalSupplies;
+
+    // Mapping from owner to owned token IDs for enumeration.
+    mapping(address => EnumerableSet.UintSet) private _ownedTokens;
+
+    /**
+     * @dev Returns the total amount of tokens stored by the contract.
+     */
+    function totalSupply() public view virtual override returns (uint256) {
+        return _allTokens.length();
+    }
+
+    /**
+     * @dev Returns the `index`-th tokenId stored by the contract.
+     */
+    function tokenByIndex(uint256 index) public view virtual override returns (uint256) {
+        return _allTokens.at(index);
+    }
+
+    /**
+     * @dev Returns the total amount of tokens for the specified `tokenId`.
+     */
+    function totalSupply(uint256 tokenId) public view virtual override returns (uint256) {
+        return _totalSupplies[tokenId];
+    }
+
+    /**
+     * @dev Indicates whether the specified `tokenId` exists or not.
+     */
+    function exists(uint256 tokenId) public view virtual override returns (bool) {
+        return _totalSupplies[tokenId] > 0;
+    }
+
+    /**
+     * @dev Returns the number of token ids held by `owner`.
+     */
+    function tokenCountOf(address owner) public view virtual override returns (uint256) {
+        return _ownedTokens[owner].length();
+    }
+
+    /**
+     * @dev Returns the `index`-th tokenId held by `owner`.
+     */
+    function tokenOfOwnerByIndex(address owner, uint256 index) public view virtual override returns (uint256) {
+        return _ownedTokens[owner].at(index);
+    }
+
+    /**
+     * @dev Update enumerability data in hook.
+     */
+    function _beforeTokenTransfer(
+        address operator,
+        address from,
+        address to,
+        uint256[] memory ids,
+        uint256[] memory amounts,
+        bytes memory data
+    )
+        internal
+        virtual
+        override
+    {
+        super._beforeTokenTransfer(operator, from, to, ids, amounts, data);
+
+        for (uint256 i = 0; i < ids.length; ++i) {
+            uint256 tokenId = ids[i];
+            uint256 amount = amounts[i];
+
+            // do nothing if amount == 0 or transfer to self
+            if (amount == 0 || from == to) {
+                continue;
+            }
+
+            // handle for the from address
+            if (from == address(0)) {
+                // mint token
+                _allTokens.add(tokenId);
```

Comment:
I think we can save a storage read here if we only insert `tokenId` if `totalSupplies[tokenId]` was `0` before the transfer.

Reply:
In fact, the OZ lib `EnumerableSet` will check the existence before inserting. Both way will cost a storage read opcode.

Reply:
```solidity
if (_totalSupplies[tokenId] == 0) {
    _allTokens.add(tokenId);
}
```

In this case `EnumerableSet` will not need to access that storage slot unless it's a new token.

Reply:
Then, it will access the `_totalSupplies` storage slot, any difference? In my mind, the gas cost of either way is ok, but the `EnumerableSet` could simplify the code. Generally, we do not write code about `set` as following:

```
if (set.contains(x)) {
    set.add(x)
}
```

Reply:
`_totalSupplies` will be accessed anyway because of the next line (`_totalSupplies[tokenId] += amount;`). This way we could have 1 storage read instead of 2 for basically all mints except for the first one. Since this is a library contract I think it makes sense to make it optimal, but I'll leave it up to you if you want to make this change or leave it as it is.

Reply:
Actually, we'd probably need to define a variable `uint256 supply = _totalSupplies[tokenId]` to have one `SLOAD` instead of two, or rely on the optimizer. This optimization can be left for a future version of this library.